base/math.jl

```diff
@@ -503,7 +504,45 @@ end
 
 Compute the hypotenuse ``\\sqrt{\\sum x_i^2}`` avoiding overflow and underflow.
 """
-hypot(x::Number...) = sqrt(sum(abs2(y) for y in x))
+hypot(x::Number...) = hypot(promote(x...)...)
+function hypot(x::T...) where T<:Number
+
+    # compute infnorm x (modeled on generic_vecnormMinusInf(x) in LinearAlgebra/generic.gl)
+    (v, s) = iterate(x)::Tuple
+    maxabs = abs(v)
+    while true
+        y = iterate(x, s)
+        y === nothing && break
+        (v, s) = y
+        vnorm = abs(v)
+        maxabs = ifelse(isnan(maxabs) | (maxabs > vnorm), maxabs, vnorm)
+    end
+    maxabsf = float(maxabs)
+
+    # compute vecnorm2(x) (modeled on generic_vecnorm2(x) in LinearAlgebra/generic.gl)
+    (maxabsf == 0 || isinf(maxabsf)) && return maxabsf
+    (v, s) = iterate(x)::Tuple
+    Tfloat = typeof(maxabsf)
+    if isfinite(length(x)*maxabsf*maxabsf) && maxabsf*maxabsf != 0 # Scaling not necessary
+        sum::promote_type(Float64, Tfloat) = abs2(v)
+        while true
+            y = iterate(x, s)
+            y === nothing && break
+            (v, s) = y
+            sum += abs2(v)
+        end
+        return convert(Tfloat, sqrt(sum))
+    else
+        sum = (abs(v)/maxabsf)^2
+        while true
+            y = iterate(x, s)
+            y === nothing && break
+            (v, s) = y
+            sum += (abs(v)/maxabsf)^2
+        end
+        return convert(Tfloat, maxabsf*sqrt(sum))
+    end
```

Comment:
Similarly, why not just:
```
return oftype(sqrt(sum(x -> (abs(x)/maxabsf)^2, x)), maxabsf)
```

Reply:
I followed the structure of `generic_vecnorm2` from `generic.jl` in `LinearAlgebra`, inlining the call to `vecnormInf` on the way. I take it your compact code is as efficient. Let me pull it apart to understand it better and verify with some benchmarking.
https://github.com/JuliaLang/julia/blob/1d3de4c3a4979d1725a42305e0c7692a15a56ea1/stdlib/LinearAlgebra/src/generic.jl#L298-L322

Reply:
Don't benchmark calls involving a splatted global.  Try
```
tst_hypot(x) = hypot(x...)
@btime tst_hypot($x)
```
etcetera

Reply:
`hypot_callsum` is better in terms of code size and maintainability, but it seems to me that `hypot_unrolled` (which matches the current PR for `base/math.jl`'s `hypot`) is more in tune with Julia's aim of optimizing speed and accuracy simultaneously. 

An alternative, if no one likes adding this function to `base`, is to remove it and require `LinearAlgebra` for variable-argument `hypot` calls. Then this function could be implemented as a call to `vecnorm2`, as it was prior to #27141.